lib/haskell/natural4/src/LS/Types.hs

```diff
@@ -214,43 +217,53 @@ type HornClause2 = HornClause BoolStructR
 
 -- | Though this is still quite 'syntactic'; would be even better if it were repns for a specific semantics
 data SimpleHlike a b c =
-  MkSimpleHL { shcGiven :: a
-             , shcRet :: b
-             , baseHL :: BaseHL
-             , shcSrcRef :: SrcRef
-             -- ^ may want to parametrize this
-             , shRLabel :: c
+  MkSimpleHL { shClauseGiven :: a
+             -- ^ The 'GIVEN' part of a horn-like rule.
+             , shClauseReturnVars :: b
+             -- ^ The 'GIVETH' part of a horn-like rule.
+             , shRuleSrcRef :: SrcRef
+             -- ^ Source reference to where the rule was defined.
+             -- While there is 'SrcRef' is usually generated by the
+             -- csv file parser, there is nothing that ties it inherently into
+             -- the csv pipeline.
+             --
+             -- TODO: Perhaps we want to be able to parameterise the type of
+             -- source reference we might have. For example, to have Source Spans
+             -- instead of only Source References.
+             , shBaseClauses :: BaseHL
+             -- ^ Clauses of the horn-like rule. Contains at least one clause.
+             , shRuleLabel :: c
+             -- ^ The label of the rule.
              }
   deriving stock (Eq, Show, Generic)
 
--- TODO: Meng pointed out that AtomicHC can be thought of as a special case of MultiClauseHL too. Need to think more abt this.
-data BaseHL = OneClause AtomicHC | MultiClause MultiClauseHL
-  deriving stock (Eq, Show, Generic)
-
-data HnBodHC =
-  MkHnBHC { hbHead :: RelationalPredicate
-          , hbBody :: BoolStructR
-          }
-  deriving stock (Eq, Show, Generic)
 
-data AtomicHC = HeadOnly HeadOnlyHC
-              | HeadAndBody HnBodHC
-  deriving stock (Eq, Show, Generic)
+newtype BaseHL = BaseHL (NonEmpty AtomicHC)
+  deriving stock (Generic)
+  deriving newtype (Eq, Show)
 
-newtype MultiClauseHL = MkMultiClauseHL (NonEmpty AtomicHC)
-  deriving stock (Show)
-  deriving newtype (Eq)
-   -- ^ MultiClause is meant to be '*2* or more' clauses
+{-# COMPLETE OneClause, MultiClause #-}
 
-_MkMultiClauseHL :: Iso' MultiClauseHL (NonEmpty AtomicHC)
-_MkMultiClauseHL = coerced
+-- | 'BaseHL' consists of exactly one clause.
+pattern OneClause :: AtomicHC -> BaseHL
```

Comment:
As the person who was responsible for the original OneClause vs MultiClause thing, I just wanted to flag that, *if* (and only if) you really want to continue working with the current L4 Rule data types, it might be worth thinking about whether this OneClause vs MultiClause taxonomy really is the way to go here. I recall not being satisfied with it, but just going with it anyway because I was very pressed for time; there may be cleaner ways to represent the data here.

Reply:
I didn't feel like the code had to work-around shortcoming of this type, do you have anything in mind in particular why might want to change it?

Reply:
If the question is, why we needed a OneClause vs MultiClause thing in the first place, it's because it seemed helpful for some of the constructs Meng wanted (I might be mis-remembering, but I think the one vs multiple clauses distinction basically seemed like a way to distinguish between certain constructs).

If the question is, why might the One vs MultiClause thing seem unsatisfactory: I can't remember the specifics now; it was just a part of the code that I remember feeling uncertain about. We don't need to get into this for now if you don't currently see any issues with it.